Library/RadarIO.Xamarin.iOS/RadarSDKImpl.cs

```diff
@@ -88,7 +88,14 @@ public Task<(RadarStatus, Location, IEnumerable<RadarEvent>, RadarUser)> TrackOn
             var src = new TaskCompletionSource<(RadarStatus, Location, IEnumerable<RadarEvent>, RadarUser)>();
             iOSBinding.Radar.TrackOnceWithCompletionHandler((status, location, ev, user) =>
             {
-                src.SetResult((status.ToSDK(), location?.ToSDK(), ev?.Select(Conversion.ToSDK).ToArray(), user?.ToSDK()));
```

Comment:
Has this been tested? Like, did we check that if the sdk throws an exception it gets caught in these catch blocks? I would assume the try blocks need to surround the sdk call (`iOSBinding.Radar.TrackOnceWithCompletionHandler(...)` as opposed to the setting of the result on the task completion source.

Reply:
@cameron-morrow-toptal Yes, this was tested.  Without this change, our app crashes if we start a trip on a geofence that doesn't exist.  Wrapping Radar.StartTrip in try/catch has no effect.  With this change, we are able to catch the exception and avoid the crash.  You can use the following code to reproduce this.  Without this PR, this snippet will crash and it will never enter the catch block.  With the change, you can see the exception message in the console and put a breakpoint in the catch section.

Here's a blog post with more on this:
https://deaddesk.top/don't-fall-for-TaskCompletionSource-traps/

```
try
            {
                await Radar.StartTrip(new RadarTripOptions
                {
                    ExternalId = "9999",
                    DestinationGeofenceTag = "testing",
                    DestinationGeofenceExternalId = "id",
                    Mode = RadarRouteMode.Car,
                    Metadata = new JSONObject
                        {
                            { "bing", "bong" },
                            { "ding", "dong" },
                            { "int", 10 },
                            { "bool", true }
                        }
                });
                Radar.StartTracking(Radar.TrackingOptionsContinuous);
            }
            catch (Exception ex)
            {
                Console.WriteLine("Exception: {0}", ex.Message);
            }
```

Reply:
The reason the try catch needs to be inside the callback handler rather than outside is the callback is running in a different context, so an exception thrown there doesn't come back up to TrackOnceWithCompletionHandler.  So, we need to catch exceptions within the callback and then set it on the TaskCompletionSource which will then put the exception back in the original call stack and this will allow us to catch it.

Reply:
You're right, good catch! I tested it and found the exception being caught was a nullref from the conversion code trying to convert a null value. Preferably it would just return the null value, alongside the informative status code, instead of this unhelpful nullref. To this end I added a few dozen null guards. 

Can you give me push permissions for this fork?

Reply:
@cameron-morrow-toptal I invited you to the repo.   I didn't want to make too many changes to the repo to modify any classes or return types, I just wanted to fix the crash.  :)

Typically, this is what we do when converting native callbacks to async/await.  Added an "error" parameter and RadarNativeException type to the lambda as an example.

```
iOSBinding.Radar.StartTripWithOptions(options.ToBinding(), (status, trip, events, error) =>
{
    src.SetResult((status.ToSDK(), trip.ToSDK(), events?.Select(Conversion.ToSDK)));
    try
    {
        if (error == null)
        {
            src.SetResult((status.ToSDK(), trip.ToSDK(), events?.Select(Conversion.ToSDK)));
        }
        else
        {
            src.SetException(new RadarNativeException(error)); //RadarNativeException would interpret the native error and expose error details as exception properties
        }
    }
    catch (Exception ex)
    {
        src.SetException(ex);
    }
});
```

Reply:
Ok gotcha :)

Seems like a good approach, although here the exception is actually coming from this call: `trip.ToSDK()` where my conversion code in the wrapper didn't expect the sdk to return a null. It seems like for the most part the sdk doesn't throw exceptions but rather returns non-success status codes. Either way, it's a problem when the exceptions are coming from the xamarin wrapper :(

Reply:
Thanks @cameron-morrow-toptal!